cobbler/items/system.py

```diff
@@ -324,10 +327,23 @@ def __init__(self, api: "CobblerAPI", *args: Any, **kwargs: Any) -> None:
         self._static = False
         self._static_routes: List[str] = []
         self._virt_bridge = enums.VALUE_INHERITED
+        self.__system_name = system_name
```

Comment:
We need to get rid of this property as it would introduce a reverse relationship. The whole reason we want indices is to make the search for different attributes faster. I will try to do that. Furthermore, I would like to prevent a change in usage for the constructor.

Reply:
We need to find an object by the value of an attribute in another object. When using a direct dependency between a system and an interface, we now do this by iterating over all systems in a loop and interfaces in a nested loop. Purely theoretically, we cannot achieve a search time independent of the number of elements without introducing an reverse relationship between objects.
The fastest search method we can get is through the associative array `collection.listing` (primary key). If we take its search time as one, the proposed method of searching for other attributes using secondary indexes has a time of two, regardless of the number of elements.
The secondary index itself represents the reverse relationship between objects:
`interface.attribute(secondary key) -> system.name (primary key) -> system`

We could get a faster search by replacing:
`self.__system_name = system_name`
to
`self.__system = system`

But then we would get a less reliable system in which the index could potentially contain references to already deleted objects.
Reverse relationship and the difficulties of maintaining it are simply a price to pay for speed.

Reply:
So after giving this a good night of sleep I sadly have to agree with you. The pity is that in #3440 we will be able to remove this as there - according to the current plan - `NetworkInterface` will become a dedicated collection. This would mean that our four problematic properties then are not sub-properties of the `System` type anymore but could be searched like any other object (with the fun part that `name` as a property is not unique anymore since multiple interfaces could be called `eth0` or `bmc`).

Reply:
At first glance, the #3440 will require only minor code edits; fundamentally little will change.